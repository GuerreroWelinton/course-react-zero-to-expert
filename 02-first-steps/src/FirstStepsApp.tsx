// Functional Component

import { ItemCounter } from './shopping-cart/ItemCounter';

interface ItemInCart {
  id: number;
  itemName: string;
  quantity: number;
}

const itemsInCart: ItemInCart[] = [
  { id: 1, itemName: 'PlayStation 4', quantity: 2 },
  { id: 2, itemName: 'XboxSeries', quantity: 3 },
  { id: 3, itemName: 'Nintendo Switch', quantity: 5 },
  { id: 4, itemName: 'GTA San Andrea', quantity: 7 },
];

export function FirstStepsApp() {
  return (
    <>
      <h1>Shopping Cart</h1>

      {itemsInCart.map(({ id, itemName, quantity }) => (
        <ItemCounter key={id} productName={itemName} quantity={quantity} />
      ))}
    </>
  );
}

// No usar el índece como key en arrays, es mala práctica
// La key debe ser un valor único
