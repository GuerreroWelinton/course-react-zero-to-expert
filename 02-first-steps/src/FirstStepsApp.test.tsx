/* eslint-disable @typescript-eslint/no-unused-vars */
import { render, screen } from '@testing-library/react';
import { afterEach, describe, expect, test, vi } from 'vitest';
import { FirstStepsApp } from './FirstStepsApp';

const mockComponent = vi.fn((_props: unknown) => <div data-testid="ItemCounter"></div>);

vi.mock('./shopping-cart/ItemCounter', () => ({
  ItemCounter: (props: unknown) => mockComponent(props),
}));

// vi.mock('./shopping-cart/ItemCounter', () => ({
//   ItemCounter: (props: { name: string; quantity: number }) => (
//     <div data-testid="ItemCounter" name={props.name} quantity={props.quantity}></div>
//   ),
// }));

describe('FirstStepsApp', () => {
  afterEach(() => {
    vi.clearAllMocks();
    // Limpiar los mock components despuÃ©s de cada prueba
  });

  test('should match snapshot', () => {
    const { container } = render(<FirstStepsApp />);

    expect(container).toMatchSnapshot();
  });

  test('should render the correct number of ItemCounter components', () => {
    render(<FirstStepsApp />);

    const itemCounterComponents = screen.getAllByTestId('ItemCounter');

    expect(itemCounterComponents.length).toBe(4);
  });

  test('should render ItemCounter with correct props', () => {
    render(<FirstStepsApp />);

    expect(mockComponent).toHaveBeenCalledTimes(4);
    expect(mockComponent).toHaveBeenCalledWith({ name: 'PlayStation 4', quantity: 2 });
    expect(mockComponent).toHaveBeenCalledWith({ name: 'XboxSeries', quantity: 3 });
    expect(mockComponent).toHaveBeenCalledWith({ name: 'Nintendo Switch', quantity: 5 });
    expect(mockComponent).toHaveBeenCalledWith({ name: 'GTA San Andrea', quantity: 7 });
  });
});

// Debo tener en cuenta que cada test que renderiza FirstStepsApp a su vez manda renderizar a ItemCounter
// la cantidad de veces del array itemsInCart(4)
