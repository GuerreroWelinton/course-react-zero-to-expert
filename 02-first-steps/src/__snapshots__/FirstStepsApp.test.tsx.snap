// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepsApp > should match snapshot 1`] = `
<div>
  <h1>
    Shopping Cart
  </h1>
  <section
    class="_itemRow_8b32a2"
  >
    <span
      class="_item-text_8b32a2"
      style="color: black;"
    >
      PlayStation 4
    </span>
    <button>
      +1
    </button>
    <span>
      2
    </span>
    <button>
      -1
    </button>
  </section>
  <section
    class="_itemRow_8b32a2"
  >
    <span
      class="_item-text_8b32a2"
      style="color: black;"
    >
      XboxSeries
    </span>
    <button>
      +1
    </button>
    <span>
      3
    </span>
    <button>
      -1
    </button>
  </section>
  <section
    class="_itemRow_8b32a2"
  >
    <span
      class="_item-text_8b32a2"
      style="color: black;"
    >
      Nintendo Switch
    </span>
    <button>
      +1
    </button>
    <span>
      5
    </span>
    <button>
      -1
    </button>
  </section>
  <section
    class="_itemRow_8b32a2"
  >
    <span
      class="_item-text_8b32a2"
      style="color: black;"
    >
      GTA San Andrea
    </span>
    <button>
      +1
    </button>
    <span>
      7
    </span>
    <button>
      -1
    </button>
  </section>
</div>
`;
